### PR TITLE
New stats

### PR DESCRIPTION
**Pull request by SergeyGaluzo**
Fixes https://microsofthealth.visualstudio.com/Health/_workitems/edit/113605/

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs [hunk 1]
```diff
@@ -117,6 +118,14 @@ internal class SqlServerSearchService : SearchService
             _schemaInformation = schemaInformation;
             _requestContextAccessor = requestContextAccessor;
             _compressedRawResourceConverter = compressedRawResourceConverter;
+
+            if (_resourceSearchParamStats == null)
+            {
+                lock (_locker)
+                {
+                    _resourceSearchParamStats ??= new ResourceSearchParamStats(_sqlRetryService, _logger, _model);
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:126]: Should it check that the stats are still null after acquiring the lock?

Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:126]: Not sure what conditions could lead to that. As long as constructor succeeds it is not null...

Comment (fhibf) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:127]: I believe the situation RB is suggesting is: 
- Two threads reach line 123 at the same time. 
- First one of the two threads reach line 125 first. 
- Second thread waits for the lock.
- First thread finishes the work.
- Second thread creates another instance of ResourceSearchParamStats.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs [hunk 2]
```diff
@@ -1108,5 +1120,198 @@ public void PopulateCommand(SqlCommand sqlCommand, short resourceTypeId, long st
                 _up.AddParameter(sqlCommand.Parameters, up);
             }
         }
+
+        private class ResourceSearchParamStats
+        {
+            private ISqlRetryService _sqlRetryService;
```
Comment (fhibf) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:1126]: Should this attribute be readonly? I only see it being initialized in the constructor.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs (hunk 1 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:127]: I missed you were using a null check assignment operator.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs [hunk 3]
```diff
@@ -1108,5 +1120,198 @@ public void PopulateCommand(SqlCommand sqlCommand, short resourceTypeId, long st
...
+            private readonly ILogger<SqlServerSearchService> _logger;
+            private readonly ISqlServerFhirModel _model;
+            private ConcurrentDictionary<(string TableName, string ColumnName, short ResourceTypeId, short SearchParamId), bool> _stats;
```
Comment (fhibf) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:1129]: As Init() is only called once, _stats can also be marked as readonly, if its initialization is moved out of Init().

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs (hunk 1 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs:127]: Yes... _resourceSearchParamStats ??=

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs (hunk 3 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs]: Done.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs (hunk 2 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs]: done

File: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerCreateStatsTests.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                _output.WriteLine($"database {stat}");
+            }
+
+            Assert.Single(statsFromCache.Where(_ => _.TableName == VLatest.TokenSearchParam.TableName
+                                        && _.ColumnName == "Code"
+                                        && _.ResourceTypeId == _sqlSearchService.Model.GetResourceTypeId(resourceType)
+                                        && _.SearchParamId == _sqlSearchService.Model.GetSearchParamId(new Uri("http://hl7.org/fhir/SearchParameter/clinical-identifier"))));
+        }
+
+        ////[Fact]
```
Comment (fhibf) [inline test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerCreateStatsTests.cs:70]: Do you want to keep these tests commented?

**Review by fhibf — APPROVED** [review]

File: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerCreateStatsTests.cs (hunk 4 above)
Comment (SergeyGaluzo) [reply test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerCreateStatsTests.cs]: I was planning to replace with new tests with unique queries. done.

**Review by fhibf — APPROVED** [review]

**Pull request merged**